File: src/lib/ingest/languages.ts
/**
 * Language detection by extension / filename, and per-repo language statistics.
 * The map is intentionally small (no linguist); colours follow GitHub's linguist palette.
 */
import type { FileInfo, LanguageStat } from '../data/schema';

export interface LanguageDef {
  name: string;
  color: string | null;
  /** Documentation / data / config languages are listed per file but excluded from stats. */
  prose?: boolean;
}

const L = (name: string, color: string | null, prose = false): LanguageDef => ({ name, color, prose });

const LANGS = {
  TypeScript: L('TypeScript', '#3178c6'),
  JavaScript: L('JavaScript', '#f1e05a'),
  Python: L('Python', '#3572a5'),
  Go: L('Go', '#00add8'),
  Rust: L('Rust', '#dea584'),
  Svelte: L('Svelte', '#ff3e00'),
  Astro: L('Astro', '#ff5a03'),
  Vue: L('Vue', '#41b883'),
  HTML: L('HTML', '#e34c26'),
  CSS: L('CSS', '#663399'),
  SCSS: L('SCSS', '#c6538c'),
  Sass: L('Sass', '#a53b70'),
  Less: L('Less', '#1d365d'),
  Shell: L('Shell', '#89e051'),
  PowerShell: L('PowerShell', '#012456'),
  Batchfile: L('Batchfile', '#c1f12e'),
  Dockerfile: L('Dockerfile', '#384d54'),
  Makefile: L('Makefile', '#427819'),
  CMake: L('CMake', '#da3434'),
  C: L('C', '#555555'),
  'C++': L('C++', '#f34b7d'),
  'C#': L('C#', '#178600'),
  'Objective-C': L('Objective-C', '#438eff'),
  Java: L('Java', '#b07219'),
  Kotlin: L('Kotlin', '#a97bff'),
  Scala: L('Scala', '#c22d40'),
  Groovy: L('Groovy', '#4298b8'),
  Swift: L('Swift', '#f05138'),
  Ruby: L('Ruby', '#701516'),
  PHP: L('PHP', '#4f5d95'),
  Perl: L('Perl', '#0298c3'),
  Lua: L('Lua', '#000080'),
  R: L('R', '#198ce7'),
  Julia: L('Julia', '#a270ba'),
  Dart: L('Dart', '#00b4ab'),
  Elixir: L('Elixir', '#6e4a7e'),
  Erlang: L('Erlang', '#b83998'),
  Haskell: L('Haskell', '#5e5086'),
  OCaml: L('OCaml', '#ef7a08'),
  'F#': L('F#', '#b845fc'),
  Clojure: L('Clojure', '#db5855'),
  Elm: L('Elm', '#60b5cc'),
  Zig: L('Zig', '#ec915c'),
  Nim: L('Nim', '#ffc200'),
  Crystal: L('Crystal', '#000100'),
  SQL: L('SQL', '#e38c00'),
  Nix: L('Nix', '#7e7eff'),
  Terraform: L('HCL', '#844fba'),
  Protobuf: L('Protocol Buffer', null),
  GraphQL: L('GraphQL', '#e10098'),
  'Jupyter Notebook': L('Jupyter Notebook', '#da5b0b'),
  TeX: L('TeX', '#3d6117'),
  Assembly: L('Assembly', '#6e4c13'),
  WebAssembly: L('WebAssembly', '#04133b'),
  Solidity: L('Solidity', '#aa6746'),
  Vim: L('Vim Script', '#199f4b'),
  'Emacs Lisp': L('Emacs Lisp', '#c065db'),
  'Common Lisp': L('Common Lisp', '#3fb68b'),
  Scheme: L('Scheme', '#1e4aec'),
  Fortran: L('Fortran', '#4d41b1'),
  Pascal: L('Pascal', '#e3f171'),
  Ada: L('Ada', '#02f88c'),
  D: L('D', '#ba595e'),
  V: L('V', '#4f87c4'),
  Gleam: L('Gleam', '#ffaff3'),
  CoffeeScript: L('CoffeeScript', '#244776'),
  Handlebars: L('Handlebars', '#f7931e'),
  Pug: L('Pug', '#a86454'),
  MDX: L('MDX', '#fcb32c'),
  // prose / data / config — listed per file, excluded from stats
  Markdown: L('Markdown', '#083fa1', true),
  JSON: L('JSON', '#292929', true),
  YAML: L('YAML', '#cb171e', true),
  TOML: L('TOML', '#9c4221', true),
  XML: L('XML', '#0060ac', true),
  INI: L('INI', '#d1dbe0', true),
  CSV: L('CSV', null, true),
  Text: L('Text', null, true),
  reStructuredText: L('reStructuredText', '#141414', true),
  AsciiDoc: L('AsciiDoc', '#73a0c5', true),
  SVG: L('SVG', '#ff9900', true),
  Ignore: L('Ignore List', '#000000', true),
  Lockfile: L('Lockfile', null, true),
  EditorConfig: L('EditorConfig', '#fff1f2', true),
  'Git Attributes': L('Git Attributes', '#f44d27', true),
  'Git Config': L('Git Config', '#f44d27', true),
} satisfies Record<string, LanguageDef>;

type LangKey = keyof typeof LANGS;

const BY_EXT: Record<string, LangKey> = {
  ts: 'TypeScript', tsx: 'TypeScript', mts: 'TypeScript', cts: 'TypeScript',
  js: 'JavaScript', jsx: 'JavaScript', mjs: 'JavaScript', cjs: 'JavaScript',
  py: 'Python', pyi: 'Python', pyw: 'Python',
  go: 'Go',
  rs: 'Rust',
  svelte: 'Svelte',
  astro: 'Astro',
  vue: 'Vue',
  html: 'HTML', htm: 'HTML', xhtml: 'HTML',
  css: 'CSS',
  scss: 'SCSS',
  sass: 'Sass',
  less: 'Less',
  sh: 'Shell', bash: 'Shell', zsh: 'Shell', fish: 'Shell', ksh: 'Shell',
  ps1: 'PowerShell', psm1: 'PowerShell', psd1: 'PowerShell',
  bat: 'Batchfile', cmd: 'Batchfile',
  dockerfile: 'Dockerfile',
  mk: 'Makefile', mak: 'Makefile',
  cmake: 'CMake',
  c: 'C', h: 'C',
  cpp: 'C++', cc: 'C++', cxx: 'C++', 'c++': 'C++', hpp: 'C++', hh: 'C++', hxx: 'C++', inl: 'C++',
  cs: 'C#', csx: 'C#',
  m: 'Objective-C', mm: 'Objective-C',
  java: 'Java',
  kt: 'Kotlin', kts: 'Kotlin',
  scala: 'Scala', sc: 'Scala',
  groovy: 'Groovy', gradle: 'Groovy',
  swift: 'Swift',
  rb: 'Ruby', rake: 'Ruby', gemspec: 'Ruby',
  php: 'PHP', phtml: 'PHP',
  pl: 'Perl', pm: 'Perl',
  lua: 'Lua',
  r: 'R', rmd: 'R',
  jl: 'Julia',
  dart: 'Dart',
  ex: 'Elixir', exs: 'Elixir',
  erl: 'Erlang', hrl: 'Erlang',
  hs: 'Haskell', lhs: 'Haskell',
  ml: 'OCaml', mli: 'OCaml',
  fs: 'F#', fsi: 'F#', fsx: 'F#',
  clj: 'Clojure', cljs: 'Clojure', cljc: 'Clojure', edn: 'Clojure',
  elm: 'Elm',
  zig: 'Zig',
  nim: 'Nim',
  cr: 'Crystal',
  sql: 'SQL',
  nix: 'Nix',
  tf: 'Terraform', tfvars: 'Terraform', hcl: 'Terraform',
  proto: 'Protobuf',
  graphql: 'GraphQL', gql: 'GraphQL',
  ipynb: 'Jupyter Notebook',
  tex: 'TeX', sty: 'TeX', cls: 'TeX', bib: 'TeX',
  asm: 'Assembly', s: 'Assembly',
  wat: 'WebAssembly', wast: 'WebAssembly',
  sol: 'Solidity',
  vim: 'Vim',
  el: 'Emacs Lisp',
  lisp: 'Common Lisp', lsp: 'Common Lisp',
  scm: 'Scheme', ss: 'Scheme', rkt: 'Scheme',
  f: 'Fortran', f90: 'Fortran', f95: 'Fortran', f03: 'Fortran', for: 'Fortran',
  pas: 'Pascal', pp: 'Pascal',
  adb: 'Ada', ads: 'Ada',
  d: 'D',
  v: 'V',
  gleam: 'Gleam',
  coffee: 'CoffeeScript',
  hbs: 'Handlebars', handlebars: 'Handlebars',
  pug: 'Pug', jade: 'Pug',
  mdx: 'MDX',
  md: 'Markdown', markdown: 'Markdown', mdown: 'Markdown', mkd: 'Markdown',
  json: 'JSON', jsonc: 'JSON', json5: 'JSON', webmanifest: 'JSON', geojson: 'JSON',
  yml: 'YAML', yaml: 'YAML',
  toml: 'TOML',
  xml: 'XML', xsl: 'XML', xsd: 'XML', plist: 'XML', csproj: 'XML', props: 'XML', targets: 'XML',
  ini: 'INI', cfg: 'INI', conf: 'INI',
  csv: 'CSV', tsv: 'CSV',
  txt: 'Text', text: 'Text',
  rst: 'reStructuredText',
  adoc: 'AsciiDoc', asciidoc: 'AsciiDoc',
  svg: 'SVG',
  lock: 'Lockfile',
};

const BY_FILENAME: Record<string, LangKey> = {
  dockerfile: 'Dockerfile',
  containerfile: 'Dockerfile',
  makefile: 'Makefile',
  gnumakefile: 'Makefile',
  'cmakelists.txt': 'CMake',
  rakefile: 'Ruby',
  gemfile: 'Ruby',
  podfile: 'Ruby',
  brewfile: 'Ruby',
  vagrantfile: 'Ruby',
  justfile: 'Makefile',
  'package-lock.json': 'Lockfile',
  'yarn.lock': 'Lockfile',
  'pnpm-lock.yaml': 'Lockfile',
  'bun.lockb': 'Lockfile',
  'cargo.lock': 'Lockfile',
  'go.sum': 'Lockfile',
  'poetry.lock': 'Lockfile',
  'composer.lock': 'Lockfile',
  'gemfile.lock': 'Lockfile',
  'flake.lock': 'Lockfile',
  '.gitignore': 'Ignore',
  '.dockerignore': 'Ignore',
  '.npmignore': 'Ignore',
  '.prettierignore': 'Ignore',
  '.eslintignore': 'Ignore',
  '.gitattributes': 'Git Attributes',
  '.gitmodules': 'Git Config',
  '.gitconfig': 'Git Config',
  '.editorconfig': 'EditorConfig',
  '.bashrc': 'Shell',
  '.bash_profile': 'Shell',
  '.zshrc': 'Shell',
  '.profile': 'Shell',
  'go.mod': 'Text',
  license: 'Text',
  licence: 'Text',
  copying: 'Text',
  readme: 'Text',
  authors: 'Text',
  changelog: 'Text',
  contributors: 'Text',
  version: 'Text',
};

/** Path segments that mark vendored / generated content (excluded from stats). */
const VENDORED_DIRS = new Set(['node_modules', 'vendor', 'dist', 'build', '.git', 'bower_components', 'third_party', '__pycache__', '.venv', 'venv']);

function basename(p: string): string {
  const i = p.lastIndexOf('/');
  return i === -1 ? p : p.slice(i + 1);
}

/** Language definition for a path, or null when unknown. */
export function detectLanguageDef(path: string): LanguageDef | null {
  const base = basename(path);
  const lower = base.toLowerCase();
  const byName = BY_FILENAME[lower];
  if (byName) return LANGS[byName];
  // "Dockerfile.dev", "Makefile.am"
  if (lower.startsWith('dockerfile.')) return LANGS.Dockerfile;
  if (lower.startsWith('makefile.')) return LANGS.Makefile;
  const dot = lower.lastIndexOf('.');
  if (dot <= 0) return null;
  const ext = lower.slice(dot + 1);
  const byExt = BY_EXT[ext];
  return byExt ? LANGS[byExt] : null;
}

/** Language name for a path (null when unknown). */
export function detectLanguage(path: string): string | null {
  return detectLanguageDef(path)?.name ?? null;
}

/** Colour for a language name (null when unknown). */
export function languageColor(name: string): string | null {
  for (const def of Object.values(LANGS)) if (def.name === name) return def.color;
  return null;
}

/** Whether a path is in a vendored/generated location or is a minified asset. */
export function isVendoredPath(path: string): boolean {
  const segs = path.split('/');
  const name = segs[segs.length - 1] ?? '';
  if (segs.slice(0, -1).some((s) => VENDORED_DIRS.has(s.toLowerCase()))) return true;
  if (/\.min\.[a-z0-9]+$/i.test(name)) return true;
  return false;
}

/** Whether a file counts toward language stats. */
export function countsTowardStats(file: Pick<FileInfo, 'path' | 'binary' | 'language'>): boolean {
  if (file.binary || !file.language) return false;
  if (isVendoredPath(file.path)) return false;
  const def = detectLanguageDef(file.path);
  if (!def || def.prose) return false;
  return true;
}

/**
 * Bytes per language over the given files (docs/config/vendored/binary excluded), sorted
 * by bytes desc then name, with percents summing to 100 (drift folded into the largest).
 */
export function languageStats(files: Iterable<FileInfo>): LanguageStat[] {
  const bytes = new Map<string, number>();
  for (const f of files) {
    if (!countsTowardStats(f)) continue;
    bytes.set(f.language!, (bytes.get(f.language!) ?? 0) + f.size);
  }
  const total = Array.from(bytes.values()).reduce((a, b) => a + b, 0);
  if (total === 0) return [];
  const stats: LanguageStat[] = Array.from(bytes.entries())
    .map(([name, b]) => ({ name, bytes: b, percent: Math.round((b / total) * 1000) / 10, color: languageColor(name) }))
    .sort((a, b) => b.bytes - a.bytes || (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  const sum = stats.reduce((a, s) => a + s.percent, 0);
  const drift = Math.round((100 - sum) * 10) / 10;
  if (drift !== 0 && stats[0]) stats[0].percent = Math.round((stats[0].percent + drift) * 10) / 10;
  return stats;
}

File: internal/ingest/languages.go
package ingest

import (
	"math"
	"regexp"
	"sort"
	"strings"

	"frznforge/internal/model"
)

// languageDef is one entry of the language map — deliberately small (no linguist); the colours
// follow GitHub's linguist palette.
type languageDef struct {
	Name  string
	Color *string
	// Prose marks documentation, data and config languages: listed per file, excluded from
	// the per-repo statistics.
	Prose bool
}

func color(hex string) *string { return &hex }

// langs is keyed the way languages.ts keys it, so the two files can be diffed side by side.
var langs = map[string]languageDef{
	"TypeScript":       {"TypeScript", color("#3178c6"), false},
	"JavaScript":       {"JavaScript", color("#f1e05a"), false},
	"Python":           {"Python", color("#3572a5"), false},
	"Go":               {"Go", color("#00add8"), false},
	"Rust":             {"Rust", color("#dea584"), false},
	"Svelte":           {"Svelte", color("#ff3e00"), false},
	"Astro":            {"Astro", color("#ff5a03"), false},
	"Vue":              {"Vue", color("#41b883"), false},
	"HTML":             {"HTML", color("#e34c26"), false},
	"CSS":              {"CSS", color("#663399"), false},
	"SCSS":             {"SCSS", color("#c6538c"), false},
	"Sass":             {"Sass", color("#a53b70"), false},
	"Less":             {"Less", color("#1d365d"), false},
	"Shell":            {"Shell", color("#89e051"), false},
	"PowerShell":       {"PowerShell", color("#012456"), false},
	"Batchfile":        {"Batchfile", color("#c1f12e"), false},
	"Dockerfile":       {"Dockerfile", color("#384d54"), false},
	"Makefile":         {"Makefile", color("#427819"), false},
	"CMake":            {"CMake", color("#da3434"), false},
	"C":                {"C", color("#555555"), false},
	"C++":              {"C++", color("#f34b7d"), false},
	"C#":               {"C#", color("#178600"), false},
	"Objective-C":      {"Objective-C", color("#438eff"), false},
	"Java":             {"Java", color("#b07219"), false},
	"Kotlin":           {"Kotlin", color("#a97bff"), false},
	"Scala":            {"Scala", color("#c22d40"), false},
	"Groovy":           {"Groovy", color("#4298b8"), false},
	"Swift":            {"Swift", color("#f05138"), false},
	"Ruby":             {"Ruby", color("#701516"), false},
	"PHP":              {"PHP", color("#4f5d95"), false},
	"Perl":             {"Perl", color("#0298c3"), false},
	"Lua":              {"Lua", color("#000080"), false},
	"R":                {"R", color("#198ce7"), false},
	"Julia":            {"Julia", color("#a270ba"), false},
	"Dart":             {"Dart", color("#00b4ab"), false},
	"Elixir":           {"Elixir", color("#6e4a7e"), false},
	"Erlang":           {"Erlang", color("#b83998"), false},
	"Haskell":          {"Haskell", color("#5e5086"), false},
	"OCaml":            {"OCaml", color("#ef7a08"), false},
	"F#":               {"F#", color("#b845fc"), false},
	"Clojure":          {"Clojure", color("#db5855"), false},
	"Elm":              {"Elm", color("#60b5cc"), false},
	"Zig":              {"Zig", color("#ec915c"), false},
	"Nim":              {"Nim", color("#ffc200"), false},
	"Crystal":          {"Crystal", color("#000100"), false},
	"SQL":              {"SQL", color("#e38c00"), false},
	"Nix":              {"Nix", color("#7e7eff"), false},
	"Terraform":        {"HCL", color("#844fba"), false},
	"Protobuf":         {"Protocol Buffer", nil, false},
	"GraphQL":          {"GraphQL", color("#e10098"), false},
	"Jupyter Notebook": {"Jupyter Notebook", color("#da5b0b"), false},
	"TeX":              {"TeX", color("#3d6117"), false},
	"Assembly":         {"Assembly", color("#6e4c13"), false},
	"WebAssembly":      {"WebAssembly", color("#04133b"), false},
	"Solidity":         {"Solidity", color("#aa6746"), false},
	"Vim":              {"Vim Script", color("#199f4b"), false},
	"Emacs Lisp":       {"Emacs Lisp", color("#c065db"), false},
	"Common Lisp":      {"Common Lisp", color("#3fb68b"), false},
	"Scheme":           {"Scheme", color("#1e4aec"), false},
	"Fortran":          {"Fortran", color("#4d41b1"), false},
	"Pascal":           {"Pascal", color("#e3f171"), false},
	"Ada":              {"Ada", color("#02f88c"), false},
	"D":                {"D", color("#ba595e"), false},
	"V":                {"V", color("#4f87c4"), false},
	"Gleam":            {"Gleam", color("#ffaff3"), false},
	"CoffeeScript":     {"CoffeeScript", color("#244776"), false},
	"Handlebars":       {"Handlebars", color("#f7931e"), false},
	"Pug":              {"Pug", color("#a86454"), false},
	"MDX":              {"MDX", color("#fcb32c"), false},

	// prose / data / config — listed per file, excluded from stats
	"Markdown":         {"Markdown", color("#083fa1"), true},
	"JSON":             {"JSON", color("#292929"), true},
	"YAML":             {"YAML", color("#cb171e"), true},
	"TOML":             {"TOML", color("#9c4221"), true},
	"XML":              {"XML", color("#0060ac"), true},
	"INI":              {"INI", color("#d1dbe0"), true},
	"CSV":              {"CSV", nil, true},
	"Text":             {"Text", nil, true},
	"reStructuredText": {"reStructuredText", color("#141414"), true},
	"AsciiDoc":         {"AsciiDoc", color("#73a0c5"), true},
	"SVG":              {"SVG", color("#ff9900"), true},
	"Ignore":           {"Ignore List", color("#000000"), true},
	"Lockfile":         {"Lockfile", nil, true},
	"EditorConfig":     {"EditorConfig", color("#fff1f2"), true},
	"Git Attributes":   {"Git Attributes", color("#f44d27"), true},
	"Git Config":       {"Git Config", color("#f44d27"), true},
}

var langByExt = map[string]string{
	"ts": "TypeScript", "tsx": "TypeScript", "mts": "TypeScript", "cts": "TypeScript",
	"js": "JavaScript", "jsx": "JavaScript", "mjs": "JavaScript", "cjs": "JavaScript",
	"py": "Python", "pyi": "Python", "pyw": "Python",
	"go":     "Go",
	"rs":     "Rust",
	"svelte": "Svelte",
	"astro":  "Astro",
	"vue":    "Vue",
	"html":   "HTML", "htm": "HTML", "xhtml": "HTML",
	"css":  "CSS",
	"scss": "SCSS",
	"sass": "Sass",
	"less": "Less",
	"sh":   "Shell", "bash": "Shell", "zsh": "Shell", "fish": "Shell", "ksh": "Shell",
	"ps1": "PowerShell", "psm1": "PowerShell", "psd1": "PowerShell",
	"bat": "Batchfile", "cmd": "Batchfile",
	"dockerfile": "Dockerfile",
	"mk":         "Makefile", "mak": "Makefile",
	"cmake": "CMake",
	"c":     "C", "h": "C",
	"cpp": "C++", "cc": "C++", "cxx": "C++", "c++": "C++", "hpp": "C++", "hh": "C++", "hxx": "C++", "inl": "C++",
	"cs": "C#", "csx": "C#",
	"m": "Objective-C", "mm": "Objective-C",
	"java": "Java",
	"kt":   "Kotlin", "kts": "Kotlin",
	"scala": "Scala", "sc": "Scala",
	"groovy": "Groovy", "gradle": "Groovy",
	"swift": "Swift",
	"rb":    "Ruby", "rake": "Ruby", "gemspec": "Ruby",
	"php": "PHP", "phtml": "PHP",
	"pl": "Perl", "pm": "Perl",
	"lua": "Lua",
	"r":   "R", "rmd": "R",
	"jl":   "Julia",
	"dart": "Dart",
	"ex":   "Elixir", "exs": "Elixir",
	"erl": "Erlang", "hrl": "Erlang",
	"hs": "Haskell", "lhs": "Haskell",
	"ml": "OCaml", "mli": "OCaml",
	"fs": "F#", "fsi": "F#", "fsx": "F#",
	"clj": "Clojure", "cljs": "Clojure", "cljc": "Clojure", "edn": "Clojure",
	"elm": "Elm",
	"zig": "Zig",
	"nim": "Nim",
	"cr":  "Crystal",
	"sql": "SQL",
	"nix": "Nix",
	"tf":  "Terraform", "tfvars": "Terraform", "hcl": "Terraform",
	"proto":   "Protobuf",
	"graphql": "GraphQL", "gql": "GraphQL",
	"ipynb": "Jupyter Notebook",
	"tex":   "TeX", "sty": "TeX", "cls": "TeX", "bib": "TeX",
	"asm": "Assembly", "s": "Assembly",
	"wat": "WebAssembly", "wast": "WebAssembly",
	"sol":  "Solidity",
	"vim":  "Vim",
	"el":   "Emacs Lisp",
	"lisp": "Common Lisp", "lsp": "Common Lisp",
	"scm": "Scheme", "ss": "Scheme", "rkt": "Scheme",
	"f": "Fortran", "f90": "Fortran", "f95": "Fortran", "f03": "Fortran", "for": "Fortran",
	"pas": "Pascal", "pp": "Pascal",
	"adb": "Ada", "ads": "Ada",
	"d":      "D",
	"v":      "V",
	"gleam":  "Gleam",
	"coffee": "CoffeeScript",
	"hbs":    "Handlebars", "handlebars": "Handlebars",
	"pug": "Pug", "jade": "Pug",
	"mdx": "MDX",
	"md":  "Markdown", "markdown": "Markdown", "mdown": "Markdown", "mkd": "Markdown",
	"json": "JSON", "jsonc": "JSON", "json5": "JSON", "webmanifest": "JSON", "geojson": "JSON",
	"yml": "YAML", "yaml": "YAML",
	"toml": "TOML",
	"xml":  "XML", "xsl": "XML", "xsd": "XML", "plist": "XML", "csproj": "XML", "props": "XML", "targets": "XML",
	"ini": "INI", "cfg": "INI", "conf": "INI",
	"csv": "CSV", "tsv": "CSV",
	"txt": "Text", "text": "Text",
	"rst":  "reStructuredText",
	"adoc": "AsciiDoc", "asciidoc": "AsciiDoc",
	"svg":  "SVG",
	"lock": "Lockfile",
}

var langByFilename = map[string]string{
	"dockerfile":        "Dockerfile",
	"containerfile":     "Dockerfile",
	"makefile":          "Makefile",
	"gnumakefile":       "Makefile",
	"cmakelists.txt":    "CMake",
	"rakefile":          "Ruby",
	"gemfile":           "Ruby",
	"podfile":           "Ruby",
	"brewfile":          "Ruby",
	"vagrantfile":       "Ruby",
	"justfile":          "Makefile",
	"package-lock.json": "Lockfile",
	"yarn.lock":         "Lockfile",
	"pnpm-lock.yaml":    "Lockfile",
	"bun.lockb":         "Lockfile",
	"cargo.lock":        "Lockfile",
	"go.sum":            "Lockfile",
	"poetry.lock":       "Lockfile",
	"composer.lock":     "Lockfile",
	"gemfile.lock":      "Lockfile",
	"flake.lock":        "Lockfile",
	".gitignore":        "Ignore",
	".dockerignore":     "Ignore",
	".npmignore":        "Ignore",
	".prettierignore":   "Ignore",
	".eslintignore":     "Ignore",
	".gitattributes":    "Git Attributes",
	".gitmodules":       "Git Config",
	".gitconfig":        "Git Config",
	".editorconfig":     "EditorConfig",
	".bashrc":           "Shell",
	".bash_profile":     "Shell",
	".zshrc":            "Shell",
	".profile":          "Shell",
	"go.mod":            "Text",
	"license":           "Text",
	"licence":           "Text",
	"copying":           "Text",
	"readme":            "Text",
	"authors":           "Text",
	"changelog":         "Text",
	"contributors":      "Text",
	"version":           "Text",
}

// vendoredDirs are path segments that mark vendored or generated content, excluded from stats.
var vendoredDirs = map[string]bool{
	"node_modules": true, "vendor": true, "dist": true, "build": true, ".git": true,
	"bower_components": true, "third_party": true, "__pycache__": true, ".venv": true, "venv": true,
}

// minifiedRe is `/\.min\.[a-z0-9]+$/i` written out longhand.
//
// The `i` flag is spelled as explicit classes rather than Go's `(?i)`, because the two do not
// agree: Go folds Unicode, so `(?i)[a-z]` also matches U+212A KELVIN SIGN and U+017F LATIN
// SMALL LETTER LONG S, while JavaScript's non-unicode `i` deliberately refuses to canonicalise
// a non-ASCII character onto an ASCII one. `foo.min.ſ` is a silly filename, but it would be
// vendored under one implementation and counted under the other.
var minifiedRe = regexp.MustCompile(`\.[mM][iI][nN]\.[a-zA-Z0-9]+$`)

func pathBase(p string) string {
	if i := strings.LastIndexByte(p, '/'); i != -1 {
		return p[i+1:]
	}
	return p
}

// detectLanguageDef resolves a path to a language definition, or nil when unknown.
func detectLanguageDef(path string) *languageDef {
	base := pathBase(path)
	lower := strings.ToLower(base)
	if key, ok := langByFilename[lower]; ok {
		def := langs[key]
		return &def
	}
	// "Dockerfile.dev", "Makefile.am"
	if strings.HasPrefix(lower, "dockerfile.") {
		def := langs["Dockerfile"]
		return &def
	}
	if strings.HasPrefix(lower, "makefile.") {
		def := langs["Makefile"]
		return &def
	}
	dot := strings.LastIndexByte(lower, '.')
	if dot <= 0 {
		return nil
	}
	if key, ok := langByExt[lower[dot+1:]]; ok {
		def := langs[key]
		return &def
	}
	return nil
}

// DetectLanguage names the language of a path, or nil when unknown.
func DetectLanguage(path string) *string {
	def := detectLanguageDef(path)
	if def == nil {
		return nil
	}
	name := def.Name
	return &name
}

// langColors indexes langs by DISPLAY name, which is not always the key: "Terraform" is shown
// as "HCL", "Protobuf" as "Protocol Buffer", "Vim" as "Vim Script", "Ignore" as "Ignore List".
//
// The TypeScript scans Object.values(LANGS) and takes the first match, i.e. declaration order.
// A Go map has no declaration order, so on the (currently impossible) collision this keeps the
// entry with the smallest key rather than whichever the runtime happened to visit first —
// languageColor must not be able to return different bytes on two runs of the same build.
// TestLanguageNamesAreUnique holds the "currently impossible" part.
var langColors = func() map[string]*string {
	out := make(map[string]*string, len(langs))
	winner := make(map[string]string, len(langs))
	for key, def := range langs {
		if held, taken := winner[def.Name]; taken && held <= key {
			continue
		}
		winner[def.Name] = key
		out[def.Name] = def.Color
	}
	return out
}()

// languageColor is the bar colour for a language name (nil when the map has none, and nil for
// the handful of languages that are deliberately colourless — the UI falls back to a neutral).
func languageColor(name string) *string {
	return langColors[name]
}

// IsVendoredPath reports whether a path sits in a vendored or generated location, or is a
// minified asset. Both language stats and the insights code-size series exclude these.
func IsVendoredPath(path string) bool {
	segs := strings.Split(path, "/")
	for _, s := range segs[:len(segs)-1] {
		if vendoredDirs[strings.ToLower(s)] {
			return true
		}
	}
	return minifiedRe.MatchString(segs[len(segs)-1])
}

// countsTowardStats reports whether a file contributes to the language breakdown.
func countsTowardStats(f model.FileInfo) bool {
	if f.Binary || f.Language == nil {
		return false
	}
	if IsVendoredPath(f.Path) {
		return false
	}
	def := detectLanguageDef(f.Path)
	return def != nil && !def.Prose
}

// LanguageStats computes bytes per language over the given files (docs, config, vendored and
// binary excluded), sorted by bytes descending then name, with percents summing to 100 —
// rounding drift is folded into the largest entry so the bars always fill the row.
func LanguageStats(files map[string]model.FileInfo) []model.LanguageStat {
	bytesPer := map[string]int64{}
	// Iterated in path order rather than map order: the sums are integers so the order cannot
	// change them, but a deterministic walk keeps this debuggable.
	for _, path := range sortedKeys(files) {
		f := files[path]
		if !countsTowardStats(f) {
			continue
		}
		bytesPer[*f.Language] += f.Size
	}
	// Never `return nil` on the empty path: the artifact says `[]`, and a nil slice marshals
	// as `null`.
	stats := []model.LanguageStat{}
	names := sortedKeys(bytesPer)
	var total int64
	for _, name := range names {
		total += bytesPer[name]
	}
	if total == 0 {
		return stats
	}
	for _, name := range names {
		b := bytesPer[name]
		stats = append(stats, model.LanguageStat{
			Name:  name,
			Bytes: b,
			// Math.round(x) and math.Round(x) part company on exact .5 for NEGATIVE x only
			// (JavaScript rounds toward +∞, Go away from zero). b and total are both
			// non-negative, so this — and the drift arithmetic below — agree everywhere.
			Percent: math.Round(float64(b)/float64(total)*1000) / 10,
			Color:   languageColor(name),
		})
	}
	sort.SliceStable(stats, func(i, j int) bool {
		if stats[i].Bytes != stats[j].Bytes {
			return stats[i].Bytes > stats[j].Bytes
		}
		return stats[i].Name < stats[j].Name
	})
	// Summed AFTER the sort, in the sorted order, because float addition is not associative:
	// summing in name order instead could land a different last bit and flip the drift.
	var sum float64
	for _, s := range stats {
		sum += s.Percent
	}
	drift := math.Round((100-sum)*10) / 10
	if drift != 0 && len(stats) > 0 {
		stats[0].Percent = math.Round((stats[0].Percent+drift)*10) / 10
	}
	return stats
}

// sortedKeys lists a map's keys in code-point order. Every ordered walk over a map in this
// package goes through it — Go randomises map iteration, and the artifact must not.
func sortedKeys[V any](m map[string]V) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// LanguageNames is every language name ingest can put on a file, sorted.
//
// Exported for internal/highlight's sync test, which has to check that the highlighter's map
// covers everything ingest can label — a name ingest emits that the highlighter does not know is
// a file rendered with no colour, which is how `.cfg` and `.conf` shipped uncoloured for a
// version. That test used to derive the list by parsing src/lib/ingest/languages.ts; Phase 9
// deleted it, and the test skipped from then on rather than failing, so it stopped covering
// anything at all. Reading the real map is both simpler and stronger than parsing a copy of it.
func LanguageNames() []string {
	out := make([]string, 0, len(langs))
	for _, def := range langs {
		out = append(out, def.Name)
	}
	sort.Strings(out)
	return out
}
